File: 060/001/robust_tools.go
package main

/*
关卡 1-8 · Go 版 · 工具设计进阶（对比 Python 版 robust_tools_real.py）

核心：四层防护 —— 重试 / 超时 / 幂等 / 降级

执行流程图（go run robust_tools.go）：

callWeather(北京)
  ① 幂等：查 map 缓存（加锁），命中直接返回
  ② 重试循环：
       context.WithTimeout 建带超时的 ctx
       http 请求用 NewRequestWithContext(ctx) —— 超时自动取消
       失败/超时 → 退避 → 重试
  ③ 降级：3 次都失败 → 返回兜底

Go 和 Python 最大的不同（超时）：
  Python：没有原生的「可取消函数调用」，得靠线程池 + Future + result(timeout)
  Go：context 原生支持取消 + 超时，http 请求天然吃 context，
      一个 WithTimeout 就搞定，不需要「协程池」

面试怎么讲（30 秒）：
  "Go 里做超时不靠协程池，靠 context。context.WithTimeout 建一个带截止时间的上下文，
  http.NewRequestWithContext 把请求绑上去，超时后请求自动被取消，Do 返回 DeadlineExceeded。"
*/

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"math/rand"
	"net/http"
	"sync"
	"time"
)

// 人为注入开关：0=纯真实；设成 0.7 则 70% 概率人为制造失败
const CHAOS = 0

// 真实天气接口（Open-Meteo 免费），超时靠 context
func getWeather(ctx context.Context, city string) (string, error) {
	if CHAOS > 0 && rand.Float64() < CHAOS {
		if rand.Float64() < 0.5 {
			return "", fmt.Errorf("人为注入：网络抖动")
		}
		return "", context.DeadlineExceeded // 人为注入：超时
	}

	url := "https://api.open-meteo.com/v1/forecast?latitude=39.9&longitude=116.4&current_weather=true"
	req, err := http.NewRequestWithContext(ctx, "GET", url, nil) // 绑定 ctx，超时自动取消
	if err != nil {
		return "", err
	}
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return "", err // ctx 超时时，这里返回的 err 包裹了 context.DeadlineExceeded
	}
	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		return "", fmt.Errorf("HTTP %d", resp.StatusCode)
	}
	var data struct {
		CurrentWeather struct {
			Temperature float64 `json:"temperature"`
			Windspeed   float64 `json:"windspeed"`
		} `json:"current_weather"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
		return "", err
	}
	w := data.CurrentWeather
	return fmt.Sprintf("%s 当前 %.1f°C，风速 %.1f km/h", city, w.Temperature, w.Windspeed), nil
}

// 幂等缓存：Go 的 map 不是并发安全的，要配一把锁（Python dict 靠 GIL 不用锁）
var (
	cacheMu sync.Mutex
	cache   = map[string]string{}
)

func callWeather(city string, maxRetry int, timeout time.Duration) string {
	// ① 幂等：查缓存
	cacheMu.Lock()
	if v, ok := cache[city]; ok {
		cacheMu.Unlock()
		return v + "（缓存）"
	}
	cacheMu.Unlock()

	// ② 重试 + 超时
	for attempt := 1; attempt <= maxRetry; attempt++ {
		ctx, cancel := context.WithTimeout(context.Background(), timeout)
		result, err := getWeather(ctx, city)
		cancel()

		if err == nil {
			cacheMu.Lock()
			cache[city] = result
			cacheMu.Unlock()
			return fmt.Sprintf("第 %d 次成功：%s", attempt, result)
		}
		// errors.Is 判断超时（真超时和人为注入的超时都算）
		if errors.Is(err, context.DeadlineExceeded) {
			fmt.Printf("  第 %d 次超时（>%v），重试...\n", attempt, timeout)
		} else {
			fmt.Printf("  第 %d 次失败（%v），重试...\n", attempt, err)
		}
		time.Sleep(time.Duration(attempt) * 500 * time.Millisecond) // 退避
	}

	// ③ 降级
	return fmt.Sprintf("%s 天气服务不可用，降级返回默认值", city)
}

func main() {
	fmt.Println("=== Go 版 · 四层防护天气工具 ===")
	fmt.Println("第 1 次查北京：", callWeather("北京", 3, 2*time.Second))
	fmt.Println("第 2 次查北京：", callWeather("北京", 3, 2*time.Second), " ← 应走缓存（幂等）")
}

File: 060/001/robust_tools.py
#!/usr/bin/env python3
"""
关卡 1-8 · 工具设计进阶（完整版）

核心：真实世界的工具调用不可靠——会超时、会偶发失败、会重复提交。
生产级 agent 要给工具包四层防护：
  1. 重试 retry       —— 偶发失败自动重试几次
  2. 超时 timeout     —— 工具卡住不无限等
  3. 幂等 idempotent  —— 重复请求不重复生效
  4. 降级 fallback    —— 彻底失败给个兜底，保证可用

执行流程图（python3 robust_tools.py）：

call_weather(城市)
  ① 幂等：查缓存，命中直接返回（不重复打后端）
  ② 重试循环（最多 3 次）：
       每次调 get_weather 套超时（超 2 秒判失败）
       失败/超时 → 退避 → 重试
  ③ 降级：3 次都失败 → 返回兜底结果

和前面关卡的区别：
  1-2 的工具 = 裸函数，失败就报错，agent 就崩
  1-8 的工具 = 包了四层防护，失败自动恢复

方法调用关系：
  call_weather() ──> _executor.submit(get_weather) ──> fut.result(timeout=...)
       ├─ _cache 查/写（幂等）
       └─ 循环重试（retry）+ 兜底返回（fallback）

面试怎么讲（30 秒）：
  "生产级工具不是裸函数，要包四层：重试扛偶发故障、超时防挂死、幂等防重复副作用、
  降级保证服务可用。这是 agent 稳定性的基础。"
"""

import time
import random
from concurrent.futures import ThreadPoolExecutor

# 模拟一个不稳定的天气 API：
#   30% 概率正常返回
#   40% 概率报「网络抖动」
#   30% 概率卡死（睡 5 秒，用来触发超时）
def get_weather(city):
    r = random.random()
    if r < 0.3:
        return f"{city} 晴，25度"
    elif r < 0.7:
        raise ConnectionError("网络抖动")
    else:
        time.sleep(5)
        return f"{city} 多云"

_executor = ThreadPoolExecutor(max_workers=4)  # 持久线程池，用来给调用套超时
_cache = {}                                     # 幂等缓存


def call_weather(city, max_retry=3, timeout=2.0):
    # ① 幂等：缓存命中直接返回，不重复打后端
    if city in _cache:
        return _cache[city] + "（缓存）"

    # ② 重试 + 超时
    for attempt in range(1, max_retry + 1):
        fut = _executor.submit(get_weather, city)
        try:
            result = fut.result(timeout=timeout)   # 超时抛 TimeoutError
            _cache[city] = result
            return f"第 {attempt} 次成功：{result}"
        except TimeoutError:
            print(f"  第 {attempt} 次超时（>{timeout}s），重试...")
        except Exception as e:
            print(f"  第 {attempt} 次失败（{e}），重试...")
        time.sleep(0.5 * attempt)   # 退避：越重试等越久

    # ③ 降级：重试耗尽，返回兜底
    return f"{city} 天气服务不可用，降级返回默认值"


if __name__ == "__main__":
    print("=== 包了四层防护的天气工具 ===")
    for i in range(3):
        print(f"\n第 {i+1} 次查北京：", call_weather("北京"))
    print("\n=== 幂等验证：再查一次北京，应走缓存 ===")
    print(" ", call_weather("北京"))
